### PR TITLE
datastore: Added bucket metadata and removed external ReplaceLast API

### PR DESCRIPTION
**Pull request by johan-bjareholt**
Solves #32 

The resulting data looks like this when doing a request over REST:

```
"metadata": {
    "first_starttime": "2018-01-01"
    "last_endtime": "2018-01-02"
}
```

If the bucket is empty:

```
"metadata": {
    "first_starttime": null
    "last_endtime": null
}
```

### CODE REVIEW

File: src/datastore/datastore.rs [hunk 1]
```diff
@@ -253,14 +244,31 @@ impl DatastoreWorker {
...
     fn get_stored_buckets(&mut self, conn: &Connection) -> Result <(), DatastoreError> {
-        let mut stmt = match conn.prepare("SELECT id, name, type, client, hostname, created, data FROM buckets") {
+        let mut stmt = match conn.prepare("
+            SELECT  buckets.id, buckets.name, buckets.type, buckets.client,
+                    buckets.hostname, buckets.created, buckets.data,
+                    min(events.starttime), max(events.endtime)
+            FROM buckets
+            INNER JOIN events ON buckets.id = events.bucketrow
+            GROUP BY buckets.id
+            ;") {
```
Comment (ErikBjare) [inline src/datastore/datastore.rs]: Is this query still about as fast? (I'm kinda bad at predicting performance of SQL queries)

Comment (johan-bjareholt) [reply src/datastore/datastore.rs]: Well, I'm no expert either, but these are my two cents: Firstly inner joins and group by are made to support very large data sets in real SQL servers and this should be negligible there, but if sqlite works differently I don't know. Secondly we only have a few buckets. Thirdly this query only runs once when the data store is initialized, after that the buckets are cached. Lastly I don't have any tests for this yet so I might have to modify this query if it doesn't work as expected. This PR need more tests.

Comment (johan-bjareholt) [reply src/datastore/datastore.rs]: Another point is that starttime and endtime is indexed, so to get those values should be pretty much instant.

**Comment by johan-bjareholt** [comment]
Splitted the large datastore test into multiple tests and created a proper test for first_starttime and last_endtime with reload of the database to make sure that it works both when cached and that the value is correct when reloading after a restart.

File: src/datastore/datastore.rs [hunk 2]
```diff
@@ -269,6 +288,10 @@ impl DatastoreInstance {
                 hostname: row.get(4)?,
                 created: row.get(5)?,
                 data: serde_json::from_str(&data_str).unwrap(),
+                metadata: BucketMetadata {
+                    first_starttime: opt_first_starttime,
+                    last_endtime: opt_last_endtime,
```
Comment (ErikBjare) [inline src/datastore/datastore.rs]: I don't really like the names `first_starttime` and `last_endtime`. Perhaps just `begins` and `ends`? Or a tuple (or ISO8601 formatted interval string) `interval` or `range`?

**Comment by johan-bjareholt** [comment]
@ErikBjare There, finally fixed it. Will merge tomorrow when the tests have passed.

**Pull request merged**